components/usage/pkg/db/cost_center.go

```diff
@@ -27,36 +29,136 @@ type CostCenter struct {
+			logger.Info("No existing cost center. Creating one.")
+			defaultSpendingLimit := c.cfg.ForUsers
+			if !attributionID.IsUser() {
+				defaultSpendingLimit = c.cfg.ForTeams
+			}
+			result = CostCenter{
+				ID:              attributionID,
+				CreationTime:    NewVarcharTime(time.Now()),
+				BillingStrategy: CostCenter_Other,
+				SpendingLimit:   defaultSpendingLimit,
+				NextBillingTime: NewVarcharTime(time.Now().AddDate(0, 1, 0)),
+			}
+			err := c.conn.Save(&result).Error
+			if err != nil {
+				return CostCenter{}, err
+			}
+		} else {
+			return CostCenter{}, err
+		}
+	}
+
+	return result, nil
+}
+
+func getCostCenter(ctx context.Context, conn *gorm.DB, attributionId AttributionID) (CostCenter, error) {
 	db := conn.WithContext(ctx)
 
 	var results []CostCenter
 	db = db.Where("id = ?", attributionId).Order("creationTime DESC").Limit(1).Find(&results)
 	if db.Error != nil {
-		return nil, fmt.Errorf("failed to get cost center: %w", db.Error)
+		return CostCenter{}, fmt.Errorf("failed to get cost center: %w", db.Error)
 	}
 	if len(results) == 0 {
-		return nil, CostCenterNotFound
+		return CostCenter{}, CostCenterNotFound
 	}
 	costCenter := results[0]
-	return &costCenter, nil
+	return costCenter, nil
 }
 
-func SaveCostCenter(ctx context.Context, conn *gorm.DB, costCenter *CostCenter) (*CostCenter, error) {
-	db := conn.WithContext(ctx)
-	costCenter.CreationTime = NewVarcharTime(time.Now())
-	db = db.Save(costCenter)
+func (c *CostCenterManager) SaveCostCenter(ctx context.Context, costCenter CostCenter) (CostCenter, error) {
+
+	// retrieving the existing cost center to maintain the readonly values
+	existingCostCenter, err := c.GetOrCreateCostCenter(ctx, costCenter.ID)
+	if err != nil {
+		return CostCenter{}, err
+	}
+
+	now := time.Now()
+
+	// we don't allow setting the creationTime or the nextBillingTime from outside
+	costCenter.CreationTime = existingCostCenter.CreationTime
+	costCenter.NextBillingTime = existingCostCenter.NextBillingTime
+
+	// Do we have billing strategy update?
+	if costCenter.BillingStrategy != existingCostCenter.BillingStrategy {
+		if existingCostCenter.BillingStrategy == CostCenter_Other {
+			// moving to stripe -> let's run a finalization
+			finalizationUsage, err := c.ComputeFinalization(ctx, costCenter)
+			if err != nil {
+				return CostCenter{}, err
+			}
+			err = UpdateUsage(ctx, c.conn, finalizationUsage)
+			if err != nil {
+				return CostCenter{}, err
+			}
+		}
+		c.updateNextBillingTime(&costCenter, now)
+	}
+
+	// we update the creationTime
+	costCenter.CreationTime = NewVarcharTime(now)
+	db := c.conn.Save(&costCenter)
 	if db.Error != nil {
-		return nil, fmt.Errorf("failed to save cost center: %w", db.Error)
+		return CostCenter{}, fmt.Errorf("failed to save cost center for attributionID %s: %w", costCenter.ID, db.Error)
 	}
 	return costCenter, nil
 }
+
+func (c *CostCenterManager) ComputeFinalization(ctx context.Context, costCenter CostCenter) (Usage, error) {
```

Comment:
```suggestion
func (c *CostCenterManager) ComputeFinalization(ctx context.Context, costCenter CostCenter) (Usage, error) {
```
From this signature, I'd have expected that it will also store the Usage record in the DB. Perhaps renaming to `InvoiceUsageRecordForCostCenter` would convey this better.

Additionally, why this is method accept the `costCenter CostCenter` when it only uses the `costCenter.ID`? Could we supply only the ID here?

Reply:
I generally use 'Compute...' to indicate that a function doesn't have aside-effects but just computes something. But I'm happy to change the name to make it clearer for you.
I'm passing in the costCenter because it contains information about `spendingLimit` which is likely to be used pretty soon, but we can fetch it within the function if needed.